spec/unit/defaults_spec.rb

```diff
@@ -189,4 +189,79 @@
       Puppet.initialize_settings
     end
   end
+
+  describe "the call hook for the cadir setting", unless: Puppet::Util::Platform.windows? do
+    it 'does not warn when the cadir is outside the puppet ssldir' do
+      FileUtils.mkdir_p(Puppet[:confdir])
+      # This is fun; in order to get the confdir setting from the spec_helper to
+      # not be deleted by the #unsafe_clear in Puppet::Settings#parse_config,
+      # we can code it into the confdir itself, and thus we can avoid getting
+      # that setting wiped out during Puppet.initialize_settings.
+      File.write(File.join(Puppet[:confdir], 'puppet.conf'),
+                 "cadir = /my/cool/path/for/my/cadir\n
+                  confdir = #{Puppet[:confdir]}")
+      expect(Puppet).to_not receive(:log_ca_migration_warning)
+      Puppet.initialize_settings
+      expect(Puppet[:cadir]).to eq("/my/cool/path/for/my/cadir")
+    end
+
+    it 'does warn when the cadir is inside the puppet ssldir' do
+      FileUtils.mkdir_p(Puppet[:confdir])
+      cadir_location = File.join(Puppet[:ssldir], 'still_inside_ssldir')
+      # This is fun; in order to get the confdir setting from the spec_helper to
+      # not be deleted by the #unsafe_clear in Puppet::Settings#parse_config,
+      # we can code it into the confdir itself, and thus we can avoid getting
+      # that setting wiped out during Puppet.initialize_settings.
+      File.write(File.join(Puppet[:confdir], 'puppet.conf'),
+                 "cadir = #{cadir_location}\n
+                  confdir = #{Puppet[:confdir]}")
+      expect(Puppet).to receive(:log_ca_migration_warning).twice
+      Puppet.initialize_settings
+      expect(Puppet[:cadir]).to eq(cadir_location)
+      end
+  end
+
+  describe "the default cadir", :unless => Puppet::Util::Platform.windows?  do
+    it 'defaults to the puppetserver confdir' do
+      Puppet.initialize_settings
+      expect(Puppet[:cadir]).to eq('/etc/puppetlabs/puppetserver/ca')
+    end
+
+    it 'returns nil for Windows platforms', :if => Puppet::Util::Platform.windows? do
+      Puppet.initialize_settings
+      expect(Puppet.default_cadir).to eq("")
```

Comment:
Should this be checking for `Puppet[:cadir]` instead of `Puppet.default_cadir`, like the one above? I don't really understand why things are different for Windows.. 😅

Reply:
Because you can't run the master on windows, so there is no CA dir. Also unix paths like `/etc` aren't valid on windows, so we could hard-code something different, but that would be pointless because it won't be used ever.

Reply:
gotcha. My question still stands about checking the setting vs. the `default_cadir`. Since there are separate tests for `default_cadir` below (though not one for windows).

Reply:
oh, yeah, that language is a little wrong, should be empty string instead of nil in the title.